grpc/federation/cel/url.go

```diff
@@ -0,0 +1,611 @@
+package cel
+
+import (
+	"context"
+	"net/url"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/common/types/traits"
+)
+
+const URLPackageName = "url"
+
+var (
+	URLType      = cel.ObjectType("grpc.federation.url.URL")
+	UserinfoType = cel.ObjectType("grpc.federation.url.Userinfo")
+)
+
+func (x *URL) GoURL() (url.URL, error) {
+	var user *url.Userinfo
+	if u := x.GetUser(); u != nil {
+		if u2, err := u.GoUserinfo(); err != nil {
+			return url.URL{}, err
+		} else {
+			user = u2
+		}
+	}
+
+	return url.URL{
+		Scheme:     x.GetScheme(),
+		Opaque:     x.GetOpaque(),
+		User:       user,
+		Host:       x.GetHost(),
+		Path:       x.GetPath(),
+		RawPath:    x.GetRawPath(),
+		ForceQuery: x.GetForceQuery(),
+		RawQuery:   x.GetRawQuery(),
+		Fragment:   x.GetFragment(),
+	}, nil
+}
+
+func (x *Userinfo) GoUserinfo() (*url.Userinfo, error) {
```

Comment:
According to the implementation, GoUserinfo won't return nil, but is it really correct? For example, should we return nil when both username and password are empty? The original UserInfo struct assumes the case when u (*UserInfo) is nil and returns an empty string.

Reply:
This issue was caused by an implementation mistake that occurred in the past. It has been corrected now.